### PR TITLE
Move between nesting levels with arrow keys in navigate mode.

### LINKED ISSUE #20732: Iterations on Navigation Mode
Columns (and likely other nested contexts) have become traps for **navigation mode**. It's not possible to press "down" from the last block of a column to move out of the column block and into the next block.

Ideally, when navigating columns, you would also be able to press right / left arrows to move selection to the next / previous column.

image

Also, pressing escape _again_ while on a child block (paragraph inside column), we should move selection one level up.

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Split out from #22453 to simplify review process.

Fixes #20732.

Add the ability to navigate between nesting levels with the arrow keys when in navigate mode. Left Arrow goes up one level (if not already at root level), Right Arrow goes down one level (if focus is on a block containing inner blocks).

## How has this been tested?

Tested navigating around a post with multiple nested blocks with keyboard only:

1. In a post containing columns, cover, navigation or other blocks with inner blocks;
2. With a block selected, press Esc to go into navigation mode;
3. Use Up and Down Arrow keys to navigate within the same level;
4. Use Right Arrow to go into a nested level (e.g. into a columns block);
5. Use Left Arrow to go out of a nesting level (e.g. out of a columns block).

## Types of changes

 New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/writing-flow/index.js [hunk 1]
```diff
@@ -380,11 +386,28 @@ export default function WritingFlow( { children } ) {
...
+				focusedBlockUid = selectionAfterEndClientId;
+			} else if ( navigateOut ) {
+				focusedBlockUid =
+					getBlockRootClientId( selectedBlockClientId ) ||
+					selectedBlockClientId;
+			} else if ( navigateIn ) {
+				focusedBlockUid =
+					getClientIdsOfDescendants( [
+						selectedBlockClientId,
+					] )[ 0 ] || selectedBlockClientId;
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/writing-flow/index.js:407]: Nullish coalescing would be more semantically correct here.
```suggestion
					] )[ 0 ] ?? selectedBlockClientId;
```

File: packages/block-editor/src/components/writing-flow/index.js [hunk 2]
```diff
@@ -380,11 +386,28 @@ export default function WritingFlow( { children } ) {
...
+
+			let focusedBlockUid;
+			if ( navigateUp ) {
+				focusedBlockUid = selectionBeforeEndClientId;
+			} else if ( navigateDown ) {
+				focusedBlockUid = selectionAfterEndClientId;
+			} else if ( navigateOut ) {
+				focusedBlockUid =
+					getBlockRootClientId( selectedBlockClientId ) ||
+					selectedBlockClientId;
```
Comment (ZebulanStanphill) [inline packages/block-editor/src/components/writing-flow/index.js:402]:
```suggestion
					getBlockRootClientId( selectedBlockClientId ) ??
					selectedBlockClientId;
```

**Review by talldan — COMMENTED** [review]
This works really well and it's great that it's actually quite a small change.

It'd be good to see some tests in the PR, I think there are already some e2e tests for navigation mode.

I wonder if this is also an opportunity to update the spoken message when switching to navigation mode:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/actions.js#L874

**Comment by ZebulanStanphill** [comment]
Tested in a Chromium-based browser on gutenberg.run. Everything seems to be working great! End-to-end test failures appear to be unrelated. (Rebasing and restarting the tests might get them to all pass.)

I agree with @talldan that some new end-to-end tests for this behavior would be nice. This is a vital part of editor accessibility, so some safeguards against breaking it unintentionally would be really helpful.

**Comment by tellthemachines** [comment]
Updated to address feedback and added an e2e test that covers Arrow key behaviour in navigation mode.

**Review by draganescu — APPROVED** [review]
I tested this and it works very good. Let's add it!

**Comment by youknowriad** [comment]
nice work here :)

**Pull request merged**